### ISSUE TITLE
Learning rate too high -> nan parameters

### ISSUE THREAD

**Issue by ThomasHeap** [alan-ppl/alan_old#12]
The ML methods are very sensitive to the lr parameter (and I believe they get more sensitive the "deeper" the hierarchical model we want to learn). 

For instance:

```python
import torch as t
import torch.nn as nn
import alan
t.manual_seed(0)

J = 2
M = 3
N = 4
platesizes = {'plate_1': J, 'plate_2': M, 'plate_3': N}
def P(tr):
    tr('a',   alan.Normal(t.zeros(()), 1))
    tr('b',   alan.Normal(tr['a'], 1))
    tr('c',   alan.Normal(tr['b'], 1), plates='plate_1')
    tr('d',   alan.Normal(tr['c'], 1), plates='plate_2')
    tr('obs', alan.Normal(tr['d'], 0.01), plates='plate_3')

class Q(alan.AlanModule):
    def __init__(self):
        super().__init__()
        self.Na = alan.MLNormal()
        self.Nb = alan.MLNormal()
        self.Nc = alan.MLNormal({'plate_1': J})
        self.Nd = alan.MLNormal({'plate_1': J, 'plate_2': M})

    def forward(self, tr):
        tr('a',   self.Na())
        tr('b',   self.Nb())
        tr('c',   self.Nc())
        tr('d',   self.Nd())


data = alan.Model(P).sample_prior(platesizes=platesizes, varnames='obs')

K = 100
T = 40
lr = 1

t.manual_seed(0)
q = Q()
m1 = alan.Model(P, q).condition(data=data)
for i in range(T):
    sample = m1.sample_same(K, reparam=False)
    print(sample.elbo().item())
    m1.update(lr, sample)
``` 

Gives ```nan``` parameters for the scale of self.Nd(). I think this is because line:

https://github.com/alan-ppl/alan/blob/9cab7d1fae35f6d72c8340264335381cdbf77258/alan/exp_fam_mixin.py#L182

Can equal 0, which then makes:

https://github.com/alan-ppl/alan/blob/9cab7d1fae35f6d72c8340264335381cdbf77258/alan/exp_fam_mixin.py#L187

Return ```nan``` entries.

Could we stabilise this?:

```python
scale = (Ex2 - loc**2).sqrt() + 1e-8
```

**Comment by ThomasHeap** [comment]
Even with this stabilisation sometimes 

```python
Ex2 - loc**2 < 0
```

Which can result in nans. Adding the epsilon inside the square root doesn't help. I'll look into this further, but so far my attempts to fix this lead to ```inf```s in the parameters. 

This could just be too high of a learning rate, the stabilisation has increased the range of usable learning rates.

**Comment by ThomasHeap** [comment]
Closing as no longer relevant.
